src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -41,149 +42,183 @@ BilinearInterpolation(
     const size_t outRowSize,
     const size_t outColSize,
     const size_t depth):
-    inRowSize(inRowSize),
-    inColSize(inColSize),
-    outRowSize(outRowSize),
-    outColSize(outColSize),
-    depth(depth),
-    batchSize(0)
-{
-  // Nothing to do here.
-}
+  inRowSize(inRowSize),
+  inColSize(inColSize),
+  outRowSize(outRowSize),
+  outColSize(outColSize),
+  depth(depth),
+  batchSize(0)
+  {
+    // Let Po denote any output matrix point. This value at this
+    // point will be determined by four points in the padded input
+    // matrix. Let these be denoted by [P1, P2, P3, P4]. This will
+    // form a unit square. So if we know the bottom right corner then
+    // we can determine the other points.
+    // Po = P1*d1 + P2*d2 + P3*d3 + P4*d4 where di's is a constant.
+    // The di's and Pi's depend on the dimensions of input matrix
+    // and output matrix. So, we can precompute these points in the
+    // constructor and use them directly in the forward and backward
+    // functions. In bilinear interpolation we will pad the input
+    // matrix to make the calculations more simple.
+    weights.set_size(WeightSize(), 1);
+    double scaleRow = (double) inRowSize / (double) outRowSize;
+    double scaleCol = (double) inColSize / (double) outColSize;
+    coeffsPre = arma::cube(weights.memptr(), 2, 2,
+      outRowSize * outColSize, false, false);
+    indexPre = arma::cube(weights.memptr() + 4 * outRowSize * outColSize,
+      1, 2, outRowSize * outColSize, false, false);
+    temp = arma::mat(weights.memptr() + 6 * outRowSize * outColSize,
+      inRowSize + 2, inColSize + 2, false, false);
+
+    for (size_t i = 0; i < outRowSize; ++i)
+    {
+      double rOrigin = (i + 0.5) * scaleRow;
+
+      for(size_t j = 0; j < outColSize; ++j)
+      {
+
+        double cOrigin = (j + 0.5) * scaleCol;
+        size_t cEnd = (size_t) std::floor(cOrigin + 0.5);
+        size_t rEnd = (size_t) std::floor(rOrigin + 0.5);
+
+        double deltaC = 0.5 + cOrigin - cEnd;
+        double deltaR = 0.5 + rOrigin - rEnd; 
+        if (deltaR >= 1)
+          deltaR -= 1;
+
+        if (deltaC >= 1)
+          deltaC -= 1;
+
+        size_t currentIdx = j * outRowSize + i;
+        // Value of di's for each point in Output matrix.
+        coeffsPre.slice(currentIdx)(0, 0) = (1 - deltaC) * (1 - deltaR);
+        coeffsPre.slice(currentIdx)(0, 1) = (1 - deltaR) * deltaC;
+        coeffsPre.slice(currentIdx)(1, 0) = deltaR * (1 - deltaC);
+        coeffsPre.slice(currentIdx)(1, 1) = deltaR * deltaC;
+        // Bottom Right corner of the Pi's for each point in Output matrix.
+        indexPre.slice(currentIdx)(0, 0) = cEnd + 1;
+        indexPre.slice(currentIdx)(0, 1) = rEnd + 1;
```

Comment:
@zoq I am not sure that we need to store these. What do you think about this. Perhaps the `indexPre` can be removed as its not that expensive. But I am not sure about `coeffsPre`. What do you think?

Reply:
Yes

Reply:
Hm, don't think this will have a huge effect on the runtime, it's mostly in the forward step, if you think it's cleaner please feel free to adjust that.